The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7869238163/job/21467940868) [(plain)](https://github.com/rust-lang/rust/commit/6581eed949b44a1f0aa1e3d294c75e77fde38024/checks/21467940868/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=d-sonuga
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_c8b758d6-7bfc-4153-9739-0ae2d7633fde
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=bc9cc99352afadd94676f625d1a7290095c91ec7
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_c8b758d6-7bfc-4153-9739-0ae2d7633fde
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_c8b758d6-7bfc-4153-9739-0ae2d7633fde
GITHUB_TRIGGERING_ACTOR=d-sonuga
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/119928/merge
GITHUB_WORKFLOW_SHA=bc9cc99352afadd94676f625d1a7290095c91ec7
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
#12 writing image sha256:f9fe0979ac768bb263744510eed48c1d0386436631cc2ca95679fce9c649519b done
#12 naming to docker.io/library/rust-ci done
#12 DONE 10.0s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Mon Feb 12 08:50:08 UTC 2024
  network time: Mon, 12 Feb 2024 08:50:08 GMT
  network time: Mon, 12 Feb 2024 08:50:08 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'build.optimized-compiler-builtins', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---

---- [ui] tests/ui/did_you_mean/collect-without-into-iter-call.rs stdout ----
diff of stderr:

26 LL | fn process(items: impl IntoIterator<Item = String>) -> Vec<String> {
27    |                   -------------------------------- method `collect` not found for this type parameter
28 LL |     items.collect()
-    |           ^^^^^^^ `impl IntoIterator<Item = String>` is not an iterator
+    |           ^^^^^^^ method cannot be called on `impl IntoIterator<Item = String>` due to unsatisfied trait bounds
30    |
- help: call `.into_iter()` first
+    = help: items from traits can only be used if the type parameter is bounded by the trait
+ help: the following trait defines an item `collect`, perhaps you need to restrict type parameter `impl IntoIterator<Item = String>` with it:
- LL |     items.into_iter().collect()
-    |           ++++++++++++
-    |           ++++++++++++
+ LL | fn process(items: impl IntoIterator<Item = String> + Iterator) -> Vec<String> {
35 
36 error: aborting due to 3 previous errors
37 

---
--- stderr -------------------------------
error[E0599]: no method named `map` found for opaque type `impl IntoIterator<Item = i32>` in the current scope
##[error]  --> /checkout/tests/ui/did_you_mean/collect-without-into-iter-call.rs:6:29
   |
LL |     let other_items = items.map(|i| i + 1);
   |                             ^^^ `impl IntoIterator<Item = i32>` is not an iterator
   |
help: call `.into_iter()` first
Build completed unsuccessfully in 0:14:17
Build completed unsuccessfully in 0:14:17
LL |     let other_items = items.into_iter().map(|i| i + 1);

error[E0599]: no method named `collect` found for opaque type `impl IntoIterator<Item = i32>` in the current scope
##[error]  --> /checkout/tests/ui/did_you_mean/collect-without-into-iter-call.rs:8:31
   |
   |
LL |     let vec: Vec<i32> = items.collect();
   |                               ^^^^^^^ `impl IntoIterator<Item = i32>` is not an iterator
   |
help: call `.into_iter()` first
   |
LL |     let vec: Vec<i32> = items.into_iter().collect();

error[E0599]: no method named `collect` found for type parameter `impl IntoIterator<Item = String>` in the current scope
##[error]  --> /checkout/tests/ui/did_you_mean/collect-without-into-iter-call.rs:17:11
   |
   |
LL | fn process(items: impl IntoIterator<Item = String>) -> Vec<String> {
   |                   -------------------------------- method `collect` not found for this type parameter
LL |     items.collect()
   |           ^^^^^^^ method cannot be called on `impl IntoIterator<Item = String>` due to unsatisfied trait bounds
   = help: items from traits can only be used if the type parameter is bounded by the trait
   = help: items from traits can only be used if the type parameter is bounded by the trait
help: the following trait defines an item `collect`, perhaps you need to restrict type parameter `impl IntoIterator<Item = String>` with it:
   |
LL | fn process(items: impl IntoIterator<Item = String> + Iterator) -> Vec<String> {

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0599`.
```

</details>

--- a/compiler/rustc_hir_typeck/src/method/suggest.rs
+++ b/compiler/rustc_hir_typeck/src/method/suggest.rs
@@ -109,6 +109,93 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         self.autoderef(span, ty).any(|(ty, _)| matches!(ty.kind(), ty::Slice(..) | ty::Array(..)))
     }
 
+    fn impl_into_iterator_should_be_iterator(
+        &self,
+        ty: Ty<'tcx>,
+        span: Span,
+        unsatisfied_predicates: &Vec<(
+            ty::Predicate<'_>,
+            Option<ty::Predicate<'_>>,
+            Option<ObligationCause<'_>>,
+        )>,
+    ) -> bool {
+        fn predicate_bounds_generic_param<'tcx>(
+            predicate: ty::Predicate<'_>,
+            generics: &'tcx ty::Generics,
+            generic_param: &ty::GenericParamDef,
+            tcx: TyCtxt<'tcx>,
+        ) -> bool {
+            if let ty::PredicateKind::Clause(ty::ClauseKind::Trait(trait_pred)) =
+                predicate.kind().as_ref().skip_binder()
+            {
+                let ty::TraitPredicate { trait_ref: ty::TraitRef { args, .. }, .. } = trait_pred;
+                if args.is_empty() {
+                    return false;
+                }
+                let Some(arg_ty) = args[0].as_type() else {
+                    return false;
+                };
+                let ty::Param(param) = arg_ty.kind() else {
+                    return false;
+                };
+                // Is `generic_param` the same as the arg for this trait predicate?
+                generic_param.index == generics.type_param(&param, tcx).index
+            } else {
+                false
+            }
+        }
+
+        fn is_iterator_predicate(predicate: ty::Predicate<'_>, tcx: TyCtxt<'_>) -> bool {
+            if let ty::PredicateKind::Clause(ty::ClauseKind::Trait(trait_pred)) =
+                predicate.kind().as_ref().skip_binder()
+            {
+                tcx.is_diagnostic_item(sym::Iterator, trait_pred.trait_ref.def_id)
+            } else {
+                false
+            }
+        }
+
+        // Does the `ty` implement `IntoIterator`?
+        let Some(into_iterator_trait) = self.tcx.get_diagnostic_item(sym::IntoIterator) else {
+            return false;
+        };
+        let trait_ref = ty::TraitRef::new(self.tcx, into_iterator_trait, [ty]);
+        let cause = ObligationCause::new(span, self.body_id, ObligationCauseCode::MiscObligation);
+        let obligation = Obligation::new(self.tcx, cause, self.param_env, trait_ref);
+        if !self.predicate_must_hold_modulo_regions(&obligation) {
+            return false;
+        }
+
+        match ty.kind() {
+            ty::Param(param) => {
+                let generics = self.tcx.generics_of(self.body_id);
+                let generic_param = generics.type_param(&param, self.tcx);
+                for unsatisfied in unsatisfied_predicates.iter() {
+                    // The parameter implements `IntoIterator`
+                    // but it has called a method that requires it to implement `Iterator`
+                    if predicate_bounds_generic_param(
+                        unsatisfied.0,
+                        generics,
+                        generic_param,
+                        self.tcx,
+                    ) && is_iterator_predicate(unsatisfied.0, self.tcx)
+                    {
+                        return true;
+                    }
+                }
+            }
+            ty::Alias(ty::AliasKind::Opaque, _) => {
+                for unsatisfied in unsatisfied_predicates.iter() {
+                    if is_iterator_predicate(unsatisfied.0, self.tcx) {
+                        return true;
+                    }
+                }
+            }
+            _ => return false,
+        }
+        false
+    }
+
     #[instrument(level = "debug", skip(self))]
     pub fn report_method_error(
         &self,
@@ -555,6 +642,15 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     "`count` is defined on `{iterator_trait}`, which `{rcvr_ty}` does not implement"
                 ));
             }
+        } else if self.impl_into_iterator_should_be_iterator(rcvr_ty, span, unsatisfied_predicates)
+        {
+            err.span_label(span, format!("`{rcvr_ty}` is not an iterator"));
+            err.multipart_suggestion_verbose(
+                "call `.into_iter()` first",
+                vec![(span.shrink_to_lo(), format!("into_iter()."))],
+                Applicability::MaybeIncorrect,
+            );
+            return Some(err);
         } else if !unsatisfied_predicates.is_empty() && matches!(rcvr_ty.kind(), ty::Param(_)) {
             // We special case the situation where we are looking for `_` in
             // `<TypeParam as _>::method` because otherwise the machinery will look for blanket

--- a/tests/ui/did_you_mean/collect-without-into-iter-call.rs
+++ b/tests/ui/did_you_mean/collect-without-into-iter-call.rs
@@ -1,0 +1,19 @@
+// Tests that the compiler suggests an `into_iter` call when an `Iterator` method
+// is called on something that implements `IntoIterator`
+
+fn main() {
+    let items = items();
+    let other_items = items.map(|i| i + 1);
+    //~^ ERROR no method named `map` found for opaque type `impl IntoIterator<Item = i32>` in the current scope
+    let vec: Vec<i32> = items.collect();
+    //~^ ERROR no method named `collect` found for opaque type `impl IntoIterator<Item = i32>` in the current scope
+}
+
+fn items() -> impl IntoIterator<Item = i32> {
+    vec![1, 2, 3]
+}
+
+fn process(items: impl IntoIterator<Item = String>) -> Vec<String> {
+    items.collect()
+    //~^ ERROR no method named `collect` found for type parameter `impl IntoIterator<Item = String>` in the current scope
+}

--- a/tests/ui/did_you_mean/collect-without-into-iter-call.stderr
+++ b/tests/ui/did_you_mean/collect-without-into-iter-call.stderr
@@ -1,0 +1,36 @@
+error[E0599]: no method named `map` found for opaque type `impl IntoIterator<Item = i32>` in the current scope
+  --> $DIR/collect-without-into-iter-call.rs:6:29
+   |
+LL |     let other_items = items.map(|i| i + 1);
+   |                             ^^^ `impl IntoIterator<Item = i32>` is not an iterator
+   |
+help: call `.into_iter()` first
+   |
+LL |     let other_items = items.into_iter().map(|i| i + 1);
+   |                             ++++++++++++
+
+error[E0599]: no method named `collect` found for opaque type `impl IntoIterator<Item = i32>` in the current scope
+  --> $DIR/collect-without-into-iter-call.rs:8:31
+   |
+LL |     let vec: Vec<i32> = items.collect();
+   |                               ^^^^^^^ `impl IntoIterator<Item = i32>` is not an iterator
+   |
+help: call `.into_iter()` first
+   |
+LL |     let vec: Vec<i32> = items.into_iter().collect();
+   |                               ++++++++++++
+
+error[E0599]: no method named `collect` found for type parameter `impl IntoIterator<Item = String>` in the current scope
+  --> $DIR/collect-without-into-iter-call.rs:17:11
+   |
+LL |     items.collect()
+   |           ^^^^^^^ `impl IntoIterator<Item = String>` is not an iterator
+   |
+help: call `.into_iter()` first
+   |
+LL |     items.into_iter().collect()
+   |           ++++++++++++
+
+error: aborting due to 3 previous errors
+
+For more information about this error, try `rustc --explain E0599`.

